### ISSUE TITLE
Support Seek and other functions for fluidsynth 2+

### ISSUE THREAD

**Issue by TheCycoONE** [libsdl-org/SDL_mixer#519]
Fluidsynth 2 supports:
Seek: https://www.fluidsynth.org/api/group__midi__player.html#gaab5fe5b25ff513919b138d77e079c75a
Tell: https://www.fluidsynth.org/api/group__midi__player.html#gabc2d114e2a6926f4db800286b2ebb1da
Duration: https://www.fluidsynth.org/api/group__midi__player.html#gab8bec2b53e07daf2313412e4663c0fa7

When Fluidsynth 2 is used, these functions should be available.

This would fix an issue for us where Mix_PlayMusic continues from where it left off instead of the beginning of the track: https://github.com/CorsixTH/CorsixTH/issues/2328

**Comment by TheCycoONE** [comment]
While making these changes it might be worth hooking Play and Stop into Resume and Pause because in fluidsynth they are equivalent.

**Comment by slouken** [comment]
That sounds like a good idea. Do you want to submit a PR?

**Comment by TheCycoONE** [comment]
I wouldn't mind but it may be awhile before I have time. If someone else beats me to it I'm just as happy.

**Comment by joanbm** [comment]
Hmmm, there's a problem with PR #521 in that FluidSynth's ticks are not milliseconds.

For example, [with this MIDI file](https://en.wikipedia.org/wiki/File:Drum_sample.mid), `Mix_MusicDuration` will return a false duration of 3 seconds since FluidSynth's `fluid_player_get_total_ticks` will return ~3000 ticks, while the actual duration is about 8 seconds long.

Some more information can be found in this issue: https://github.com/FluidSynth/fluidsynth/issues/1151. The TL;DR is that FluidSynth deems reporting duration and seeking in milliseconds to be out-of-scope.

**Comment by sezero** [comment]
Well, just tried it, and playmus program reports wrong position and 0 duration for a midi file I tried. So there is definitely something wrong with the patch. And I don't know fluidsynth much to offer a solution. @TheCycoONE ?

**Comment by joanbm** [comment]
> Well, just tried it, and playmus program reports wrong position and 0 duration for a midi file I tried.

The '0 duration' problem may be because FluidSynth's `fluid_player_get_total_ticks` returns 0 until the playback *actually* starts. See https://github.com/FluidSynth/fluidsynth/issues/648.

A simple hack I found for this issue is that if after calling `fluid_player_play`, you call `fluid_synth_process(synth, 1, 0, NULL, 0, NULL);`, then `fluid_player_get_total_ticks` will return the right value. Though that's pretty much useless because it's measured in ticks, not milliseconds.

So far I haven' t been able to find any good way to translate ticks ↔️ milliseconds using *only* FluidSynth. As far as I can tell, it can be done by registering a callback using `fluid_player_set_playback_callback` to listen for "set tempo" events (i.e. `fluid_midi_event_get_type(event) = 0x51`), then processing the entire file using `fluid_synth_process` to get all events, but performance is not great since this triggers a lot of the logic for rendering a MIDI to an audio buffer.

For my use case I'm looking into using `libsmf` in addition to FluidSynth to deal with this problem more cleanly.
For SDL2_Mixer I think it should be possible to use/adapt the already built-in libtimidity to do it, or to try to contribute a feature upstream in FluidSynth.

**Comment by sezero** [comment]
I'm afraid it'll turn out to be way too much trouble than it's worth and we'll have to revert the patches

**Comment by slouken** [comment]
I agree, let's revert the initial commit and this can come back as a new PR once everything is ironed out.

**Comment by sezero** [comment]
Patch reverted from both SDL2 and main branches. That reverted the `FLUIDSYNTH_Pause` and `FLUIDSYNTH_Resume` additions too: if we do really want them, one should hand-pick them and put them back.

**Comment by joanbm** [comment]
Yeah, I think that the right place to do this is in FluidSynth itself. And I don't think it's *that* complicated (famous last words?), since FluidSynth already has all the necessary pieces internally.

• `Tell` can be done by exposing [this property](https://github.com/FluidSynth/fluidsynth/blob/7633d4aec77452e0184348d94ea13f13868633c2/src/midi/fluid_midi.h#L308), I think.
• `Duration` should be doable similarly to the [total ticks code](https://github.com/FluidSynth/fluidsynth/blob/7633d4aec77452e0184348d94ea13f13868633c2/src/midi/fluid_midi.c#L2580), but keeping track of the current tempo along the way by looking for `MIDI_SET_TEMPO` events.
• `Seek` can also be done with the same iteration logic as `Duration`, but breaking after you find the first event with a timestamp >= desired time. Then you also keep track of the corresponding ticks for that event and call `fluid_player_seek` on that

**Comment by TheCycoONE** [comment]
The main thing I was going for was resetting the track on Play and supporting Pause/Resume.

My understanding of https://www.fluidsynth.org/api/group__sequencer.html#ga7f48dab4cde18da6e0fa76458849c19a was that ticks defaulted to ms but I admit I didn't test that aspect.

**Comment by slouken** [comment]
> The main thing I was going for was resetting the track on Play and supporting Pause/Resume.

Feel free to resubmit a PR with just those aspects implemented.

**Comment by icculus** [comment]
Seek/tell/duration works with FluidSynth in SDL3_mixer. It needs FluidSynth version 2 or later to work, naturally.

**Comment by joanbm** [comment]
I did some tests on the current master branch and there is some basic support but it's still somewhat broken:

* **Duration:** In my tests I always get `MIX_GetAudioDuration(audio) = -1 = MIX_DURATION_UNKNOWN`.
  * [This seems to always be the case on the FluidSynth decoder](https://github.com/libsdl-org/SDL_mixer/blob/59a2f77645404aee3ab541c6c860970da7572bed/src/decoder_fluidsynth.c#L170C6-L170C7).
* **Seek:** In my tests `MIX_SetTrackPlaybackPosition(track, MIX_TrackMSToFrames(track, 24000))` will actually seek to 24000 ticks, not 24000 milliseconds. So if e.g. the MIDI's tempo is 120BPM, it will actually seek to 48000 milliseconds.
  * [This line in the decoder makes the assumption 1 tick = 1 ms](https://github.com/libsdl-org/SDL_mixer/blob/59a2f77645404aee3ab541c6c860970da7572bed/src/decoder_fluidsynth.c#L341).
* **Tell:** From my tests I think `MIX_GetTrackPlaybackPosition` actually returns the number of sample frames correctly. But note that this is inconsistent with the behaviour of seek (tick/ms confusion), so running `MIX_SetTrackPlaybackPosition(track, MIX_GetTrackPlaybackPosition(track));` will typically skip a chunk of the track, instead of doing nothing.

Probably this is hard to fix because FluidSynth can only seek to ticks, which depend on the MIDI's tempo events.

**Comment by icculus** [comment]
Foolish game developer that I am, I assumed "ticks" meant "milliseconds," which they don't here, so this _is_ a legit bug to be fixed.

**Comment by TheCycoONE** [comment]
> Foolish game developer that I am, I assumed "ticks" meant "milliseconds," which they don't here, so this _is_ a legit bug to be fixed.

I made the same mistake with my first patch. I feel in better company now 😀

**Comment by icculus** [comment]
With the current state of fluidsynth, the only way this is going to happen is to brute-force our way to the right position in the output:

```diff
diff --git a/src/decoder_fluidsynth.c b/src/decoder_fluidsynth.c
index d5bb502e..f77a6e69 100644
--- a/src/decoder_fluidsynth.c
+++ b/src/decoder_fluidsynth.c
@@ -90,6 +90,7 @@ typedef struct FLUIDSYNTH_TrackData
     fluid_synth_t *synth;
     fluid_settings_t *settings;
     fluid_player_t *player;
+    Uint64 current_frame;
     int freq;
 } FLUIDSYNTH_TrackData;
 
@@ -384,30 +385,45 @@ static bool SDLCALL FLUIDSYNTH_decode(void *track_userdata, SDL_AudioStream *str
     }
 
     SDL_PutAudioStreamData(stream, samples, sizeof (samples));
+    tdata->current_frame += SDL_arraysize(samples) / 2;
     return true;
 }
 
 static bool SDLCALL FLUIDSYNTH_seek(void *track_userdata, Uint64 frame)
 {
-#if (FLUIDSYNTH_VERSION_MAJOR < 2)
-    return SDL_Unsupported();
-#else
     FLUIDSYNTH_TrackData *tdata = (FLUIDSYNTH_TrackData *) track_userdata;
-    Sint64 ticks = MIX_FramesToMS(tdata->freq, frame);
-    if (ticks == -1) {
-        ticks = 0;
+
+    // This is expensive, but it's not trivial to seek to a specific frame in fluidsynth for various reasons.
+    //  (see some explanations in https://github.com/libsdl-org/SDL_mixer/issues/519)
+    if (tdata->current_frame > frame) {
+        if (fluidsynth.fluid_player_seek(tdata->player, 0) != FLUID_OK) {
+            return SDL_SetError("Couldn't rewind MIDI track");
+        }
+        tdata->current_frame = 0;
+
+        if (fluidsynth.fluid_player_get_status(tdata->player) != FLUID_PLAYER_PLAYING) {
+            if (fluidsynth.fluid_player_play(tdata->player) != FLUID_OK) {
+                return SDL_SetError("Failed to restart FluidSynth player");
+            }
+        }
     }
 
-    // !!! FIXME: docs say this will fail if a seek was requested and then a second seek happens before we play more of the midi file, since the first seek will still be in progress.
-    bool result = (fluidsynth.fluid_player_seek(tdata->player, (int)ticks) == FLUID_OK);
+    Uint64 remaining_frames = frame - tdata->current_frame;
+    while (remaining_frames > 0) {
+        if (fluidsynth.fluid_player_get_status(tdata->player) != FLUID_PLAYER_PLAYING) {
+            return SDL_SetError("Seek past end of MIDI file");
+        }
 
-    if (result && fluidsynth.fluid_player_get_status(tdata->player) != FLUID_PLAYER_PLAYING) {
-        /* start playing if player is done */
-        result = (fluidsynth.fluid_player_play(tdata->player) == FLUID_OK);
+        float samples[512];
+        const Uint64 write_frames = SDL_min((Uint64) (SDL_arraysize(samples) / 2), remaining_frames);
+        if (fluidsynth.fluid_synth_write_float(tdata->synth, (int) write_frames, samples, 0, 2, samples, 1, 2) != FLUID_OK) {
+            return SDL_SetError("Seek past end of MIDI file");  // maybe EOF...?
+        }
+        tdata->current_frame += write_frames;
+        remaining_frames -= write_frames;
     }
 
-    return result;
-#endif
+    return true;
 }
```

This is _stupid_ expensive (seeking 1 second into a file takes like 25 milliseconds), but produces correct results and honestly, it's going to be the same if we set up callbacks to catch tempo changes and stuff, since it's more or less going to do the same thing anyhow: generate and throw away output until we get to where we want. Maybe I'm wrong?

(Note that seeking to frame 0 is basically free, though, so looping back to start is cheap in any case. If you want to have an initial section that plays once and then loops from some middle section, I'd encourage you not to use MIDI files.)

**Comment by icculus** [comment]
I pushed this change for now. It sucks, but it gets the job done.
